new/rfc-knife-bootstrap-client-config.md

```diff
@@ -0,0 +1,138 @@
+---
+RFC: unassigned
+Author: Jay Mundrawala <jdm@chef.io>
+Author: Bryan McLellan <btm@chef.io>
+Status: Draft
+Type: Standards Track
+
+---
+
+# Knife Bootstrap `client.rb` Configuration
+
+Knife bootstrap currently only allows a select few `client.rb` parameters
+to be specified during bootstrap. This RFC proposes a solution to allow users
+to pass bulk configuration data to `knife bootstrap` to be placed in the
+`client.rb` before the first chef-client run.
+
+## Background
+
+The `knife bootstrap` command supports a small subset of possible parameters
+a user can configure in the `client.rb`. The chef-client cookbook provides a
+template that supports many more, allowing the user to configure additional
+options during a Chef client converge. If something you are trying to do does
+not fit into the template provided, you are also free to specify config files
+in a `client.d` folder that will automatically be loaded. However, because
+the chef-client cookbook is run along with the rest of the run list on the
+first run, your settings do not take effect until the second chef-client
+run.
+
+## Motivation
+
+    As a Chef user,
+    I want to be able to configure chef-client with options that aren't supported by knife bootstrap,
+    so that I do not have to maintain a fork of the bootstrap template,
+    or so that I do not have to learn about bootstrap templates at all,
+    or so that I don't have to make two chef-client runs on bootstrap to apply and use my configuration file.
+
+    As a Chef developer,
+    I want the user to have a generic way to pass custom client configuration file,
+    so that I don't have the maintenance cost of additional bootstrap arguments for new features
+
+## Specification
+
+### Append Config Files From `client.d`
+
+The `chef-client` cookbook adds a block to load all `.rb` files in
+the `client.d` folder. The proposal here is to do the same thing.
+Knife bootstrap will take a `client.d` folder, by default looking
+the chef repo under `bootstrap`, optionally overwritable with a
+switch `--client_d_dir`. The files will be loaded in alphabetical
+order. Knife will still sync all files under `client.d`, however
+only top-level files will the automatically loaded. The block to
+do this would look something like:
```

Comment:
we were talking in IRC today, the client_d directory should default to like #{Chef::Config[:config_dir]}/client.d" and should be a Chef::Config setting like Chef::Config[:client_d_dir] that you could override in your knife.rb or client.rb or whatever. but don't add a CLI arg like `--client_d_dir` unless we're convinced we need it.